modelforge/potential/utils.py

```diff
@@ -424,48 +425,99 @@ def compute_angular_sub_aev(self, vectors12: torch.Tensor) -> torch.Tensor:
 from abc import ABC, abstractmethod
 
 
-class RadialBasisFunction(ABC):
+class RadialBasisFunctionCore(ABC):
+    @staticmethod
     @abstractmethod
-    def compute(self, distances, centers, scale_factors):
+    def compute(nondimensionalized_distances: torch.Tensor) -> torch.Tensor:
+        """
+        Parameters
+        ---------
+        nondimensionalized_distances: torch.Tensor, shape [number_of_pairs, number_of_radial_basis_functions]
+            Nondimensional quantities that depend on pairwise distances.
+
+        Returns
+        ---------
+        torch.Tensor, shape [number_of_pairs, number_of_radial_basis_functions]
+        """
         pass
 
 
-class GaussianRadialBasisFunction(RadialBasisFunction):
-    def compute(
-        self,
-        distances: torch.Tensor,
-        centers: torch.Tensor,
-        scale_factors: torch.Tensor,
-    ) -> torch.Tensor:
-        diff = distances - centers
-        return torch.exp((-1 * scale_factors) * diff**2)
+class GaussianRadialBasisFunctionCore(RadialBasisFunctionCore):
 
+    @staticmethod
+    def compute(nondimensionalized_distances: torch.Tensor) -> torch.Tensor:
+        return torch.exp(-(nondimensionalized_distances ** 2))
 
-class DoubleExponentialRadialBasisFunction(RadialBasisFunction):
-    def compute(
-        self,
-        distances: torch.Tensor,
-        centers: torch.Tensor,
-        scale_factors: torch.Tensor,
-    ) -> torch.Tensor:
-        diff = distances - centers
-        return torch.exp(-torch.abs(diff / scale_factors))
 
+class RadialBasisFunction(nn.Module, ABC):
 
-class RadialSymmetryFunction(nn.Module):
     def __init__(
-        self,
-        number_of_radial_basis_functions: int,
-        max_distance: unit.Quantity,
-        min_distance: unit.Quantity = 0.0 * unit.nanometer,
-        dtype: Optional[torch.dtype] = None,
-        trainable: bool = False,
-        radial_basis_function: RadialBasisFunction = GaussianRadialBasisFunction(),
+            self,
+            radial_basis_function: Type[RadialBasisFunctionCore],
+            dtype,
+            prefactor: float = 1.0,
+            trainable_prefactor: bool = False,
     ):
-        """RadialSymmetryFunction class.
+        super().__init__()
+        if trainable_prefactor:
+            self.prefactor = nn.Parameter(torch.tensor([prefactor], dtype=dtype))
+        else:
+            self.register_buffer("prefactor", torch.tensor([prefactor], dtype=dtype))
+        self.radial_basis_function = radial_basis_function
 
-        Initializes and contains the logic for computing radial symmetry functions.
+    @abstractmethod
+    def nondimensionalize_distances(self, distances: torch.Tensor) -> torch.Tensor:
```

Comment:
I thought about this since it was annoying that the subtract-and-divide logic was repeated, but I felt it was very confusing for centers and means to have units of nanometers in some subclasses and to be unitless in other subclasses. I wanted to make it more clear that cancellation of units is actually happening via an implied multiplier with the inner exp and everything outside that is unitless. (Yuanqing had actually made this multiplier an explicit parameter called "alpha" in his implementation.)

Reply:
that makes sense! thanks for clarifying!